pkg/controller/handlers/toolreference/toolreference.go

```diff
@@ -434,11 +445,16 @@ func (h *Handler) readMCPCatalogDirectory(catalog string) ([]catalogEntryInfo, e
 	return entries, nil
 }
 
-func (h *Handler) readFromMCPCatalogs(ctx context.Context, c client.Client) error {
+func (h *Handler) readFromMCPCatalogs(ctx context.Context, c client.Client, catalogs []v1.Catalog) error {
 	var (
 		toAdd []client.Object
 		errs  []error
 	)
+
+	h.catalogLock.Lock()
+	defer h.catalogLock.Unlock()
```

Comment:
We shouldn't hold this lock during IO.
```suggestion
	h.catalogLock.Lock()
	mcpCatalogs := h.mcpCatalogs
	h.catalogLock.Unlock()
```

Reply:
Refactored, no longer needed a mutex.